File: lib/EasyInk.Net/EasyInk.Printer/src/Services/PrintDebugLogService.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using EasyInk.Engine;
using EasyInk.Engine.Models;
using EasyInk.Printer.Config;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace EasyInk.Printer.Services;

public sealed class PrintDebugLogService : IDisposable
{
    private const string RequestFileName = "request.json";
    private const string InputPdfFileName = "input.pdf";
    private const string ManifestFileName = "manifest.json";
    private const string EngineLogFileName = "engine.log";
    private const string SubmitResultFileName = "submit-result.json";
    private const string CompletionResultFileName = "completion-result.json";
    private static readonly TimeSpan TerminalSubmitGracePeriod = TimeSpan.FromHours(1);

    private readonly bool _enabled;
    private readonly string _artifactRoot;
    private readonly int _retentionCount;
    private readonly ConcurrentDictionary<string, string> _artifactPaths = new();
    private readonly System.Threading.Timer? _cleanupTimer;

    public PrintDebugLogService(HostConfig config)
    {
        _enabled = config.PrintDebugLoggingEnabled;
        _artifactRoot = HostConfig.ResolvePrintDebugArtifactsDir(config.PrintDebugArtifactsDir!);
        _retentionCount = Math.Max(1, config.PrintDebugArtifactRetentionCount);

        if (!_enabled) return;

        try
        {
            Directory.CreateDirectory(_artifactRoot);
            CleanupExpiredArtifacts();
            _cleanupTimer = new System.Threading.Timer(
                _ => CleanupExpiredArtifactsSafe(),
                null,
                TimeSpan.FromDays(1),
                TimeSpan.FromDays(1));
        }
        catch (Exception ex)
        {
            SimpleLogger.Error("打印调试日志初始化失败", ex);
        }
    }

    public bool Enabled => _enabled;

    public void BeginPrintRequest(string requestId, string command, string? parametersJson, byte[]? pdfBytes)
    {
        if (!_enabled || string.IsNullOrWhiteSpace(requestId)) return;

        try
        {
            var directory = CreateArtifactDirectory(requestId);
            _artifactPaths[requestId] = directory;

            var requestJson = BuildRequestJson(command, parametersJson, pdfBytes, directory);
            File.WriteAllText(Path.Combine(directory, RequestFileName), requestJson);

            SavePdfBytes(directory, pdfBytes, overwrite: false);
            WriteManifest(directory, requestId, command);
            CleanupExpiredArtifacts();

            SimpleLogger.Info($"打印调试附件已创建: jobId={requestId}, path={directory}");
        }
        catch (Exception ex)
        {
            SimpleLogger.Error($"打印调试附件保存失败: jobId={requestId}", ex);
        }
    }

    public void BeginPrintRequest(string requestId, string command, JObject? parameters, byte[]? pdfBytes)
    {
        BeginPrintRequest(requestId, command, parameters?.ToString(Formatting.None), pdfBytes);
    }

    public void WriteSubmitResult(string requestId, PrinterResult result)
    {
        WriteResult(requestId, SubmitResultFileName, result);
    }

    public void WriteCompletionResult(string requestId, PrintRequestParams request, PrinterResult result)
    {
        if (!_enabled) return;

        var directory = GetArtifactDirectory(requestId);
        if (directory == null) return;

        try
        {
            var payload = new
            {
                completedAt = DateTime.Now,
                request = new
                {
                    request.PrinterName,
                    request.Copies,
                    request.Dpi,
                    request.Landscape,
                    request.PaperSize,
                    request.Offset,
                    request.UserData,
                    hasPdfBytes = request.PdfBytes != null && request.PdfBytes.Length > 0,
                    hasPdfBase64 = !string.IsNullOrEmpty(request.PdfBase64),
                    hasPdfUrl = !string.IsNullOrEmpty(request.PdfUrl),
                    request.PdfUrl
                },
                result
            };
            File.WriteAllText(
                Path.Combine(directory, CompletionResultFileName),
                JsonConvert.SerializeObject(payload, Formatting.Indented, EasyInk.Engine.JsonConfig.CamelCase));
        }
        catch (Exception ex)
        {
            SimpleLogger.Error($"打印调试完成结果保存失败: jobId={requestId}", ex);
        }
    }

    public void AppendEngineLog(string? jobId, LogLevel level, string message)
    {
        if (!_enabled || string.IsNullOrWhiteSpace(jobId)) return;

        var directory = GetArtifactDirectory(jobId!);
        if (directory == null) return;

        try
        {
            var line = $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] [{level}] {message}";
            File.AppendAllText(Path.Combine(directory, EngineLogFileName), line + Environment.NewLine);
        }
        catch (Exception ex)
        {
            SimpleLogger.Error($"打印调试引擎日志保存失败: jobId={jobId}", ex);
        }
    }

    private void WriteResult(string requestId, string fileName, PrinterResult result)
    {
        if (!_enabled) return;

        var directory = GetArtifactDirectory(requestId);
        if (directory == null) return;

        try
        {
            var payload = new { writtenAt = DateTime.Now, result };
            File.WriteAllText(
                Path.Combine(directory, fileName),
                JsonConvert.SerializeObject(payload, Formatting.Indented, EasyInk.Engine.JsonConfig.CamelCase));
        }
        catch (Exception ex)
        {
            SimpleLogger.Error($"打印调试结果保存失败: jobId={requestId}", ex);
        }
    }

    private string CreateArtifactDirectory(string requestId)
    {
        var dayDirectory = Path.Combine(_artifactRoot, DateTime.Now.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture));
        Directory.CreateDirectory(dayDirectory);

        var directoryName = DateTime.Now.ToString("yyyyMMdd-HHmmssfff", CultureInfo.InvariantCulture) + "_" + SanitizeFileName(requestId);
        var directory = Path.Combine(dayDirectory, directoryName);
        Directory.CreateDirectory(directory);
        return directory;
    }

    private string? GetArtifactDirectory(string requestId)
    {
        if (_artifactPaths.TryGetValue(requestId, out var directory) && Directory.Exists(directory))
            return directory;

        var found = Directory.Exists(_artifactRoot)
            ? Directory.GetDirectories(_artifactRoot, "*", SearchOption.AllDirectories)
                .Where(d => Path.GetFileName(d).EndsWith("_" + SanitizeFileName(requestId), StringComparison.OrdinalIgnoreCase))
                .OrderByDescending(Directory.GetCreationTimeUtc)
                .FirstOrDefault()
            : null;

        if (found != null)
            _artifactPaths[requestId] = found;

        return found;
    }

    private string BuildRequestJson(string command, string? parametersJson, byte[]? pdfBytes, string directory)
    {
        JToken parameters;
        if (string.IsNullOrWhiteSpace(parametersJson))
        {
            parameters = new JObject();
        }
        else
        {
            parameters = JToken.Parse(parametersJson!);
        }

        RedactApiKey(parameters);
        ExtractPdfBase64(parameters, directory, pdfBytes != null && pdfBytes.Length > 0);

        if (pdfBytes != null && pdfBytes.Length > 0 && parameters is JObject obj)
        {
            obj["pdfBytes"] = new JObject
            {
                ["savedAs"] = InputPdfFileName,
                ["bytes"] = pdfBytes.Length,
                ["sha256"] = ComputeSha256(pdfBytes)
            };
        }

        var payload = new JObject
        {
            ["createdAt"] = DateTime.Now,
            ["command"] = command,
            ["parameters"] = parameters
        };
        return payload.ToString(Formatting.Indented);
    }

    private static void RedactApiKey(JToken token)
    {
        if (token is JObject obj)
        {
            foreach (var prop in obj.Properties().ToList())
            {
                if (string.Equals(prop.Name, "apiKey", StringComparison.OrdinalIgnoreCase) ||
                    string.Equals(prop.Name, "x-api-key", StringComparison.OrdinalIgnoreCase))
                {
                    prop.Value = "***";
                    continue;
                }

                RedactApiKey(prop.Value);
            }
        }
        else if (token is JArray array)
        {
            foreach (var item in array)
                RedactApiKey(item);
        }
    }

    private static void ExtractPdfBase64(JToken token, string directory, bool hasPdfBytes)
    {
        if (!(token is JObject obj)) return;

        var prop = obj.Properties().FirstOrDefault(p => string.Equals(p.Name, "pdfBase64", StringComparison.OrdinalIgnoreCase));
        if (prop != null && prop.Value.Type == JTokenType.String)
        {
            var value = prop.Value.ToString();
            try
            {
                var bytes = Convert.FromBase64String(value);
                if (!hasPdfBytes)
                    SavePdfBytes(directory, bytes, overwrite: true);

                prop.Value = new JObject
                {
                    ["savedAs"] = InputPdfFileName,
                    ["bytes"] = bytes.Length,
                    ["sha256"] = ComputeSha256(bytes)
                };
            }
            catch (FormatException)
            {
                prop.Value = new JObject
                {
                    ["invalidBase64"] = true,
                    ["length"] = value.Length
                };
            }
        }

        foreach (var child in obj.Properties().Select(p => p.Value).ToList())
            ExtractPdfBase64(child, directory, hasPdfBytes);
    }

    private static void SavePdfBytes(string directory, byte[]? pdfBytes, bool overwrite)
    {
        if (pdfBytes == null || pdfBytes.Length == 0) return;

        var path = Path.Combine(directory, InputPdfFileName);
        if (!overwrite && File.Exists(path)) return;
        File.WriteAllBytes(path, pdfBytes);
    }

    private static string ComputeSha256(byte[] bytes)
    {
        using var sha = SHA256.Create();
        var hash = sha.ComputeHash(bytes);
        return BitConverter.ToString(hash).Replace("-", string.Empty).ToLowerInvariant();
    }

    private static string SanitizeFileName(string value)
    {
        var invalid = Path.GetInvalidFileNameChars();
        var chars = value.Select(ch => invalid.Contains(ch) ? '_' : ch).ToArray();
        var sanitized = new string(chars).Trim();
        return string.IsNullOrEmpty(sanitized) ? Guid.NewGuid().ToString("N") : sanitized;
    }

    private static void WriteManifest(string directory, string requestId, string command)
    {
        var manifest = new
        {
            requestId,
            command,
            createdAt = DateTime.Now,
            path = directory
        };
        File.WriteAllText(Path.Combine(directory, ManifestFileName), JsonConvert.SerializeObject(manifest, Formatting.Indented));
    }

    private void CleanupExpiredArtifactsSafe()
    {
        try
        {
            CleanupExpiredArtifacts();
        }
        catch (Exception ex)
        {
            SimpleLogger.Error("打印调试附件清理失败", ex);
        }
    }

    private void CleanupExpiredArtifacts()
    {
        if (!Directory.Exists(_artifactRoot)) return;

        var artifactDirectories = Directory.GetFiles(_artifactRoot, ManifestFileName, SearchOption.AllDirectories)
            .Select(Path.GetDirectoryName)
            .Where(d => !string.IsNullOrEmpty(d))
            .Cast<string>()
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .OrderByDescending(Directory.GetCreationTimeUtc)
            .ToList();

        foreach (var directory in artifactDirectories.Skip(_retentionCount))
        {
            try
            {
                if (CanDeleteArtifactDirectory(directory))
                    Directory.Delete(directory, recursive: true);
            }
            catch
            {
            }
        }

        foreach (var dayDirectory in Directory.GetDirectories(_artifactRoot))
        {
            try
            {
                if (!Directory.EnumerateFileSystemEntries(dayDirectory).Any())
                    Directory.Delete(dayDirectory);
            }
            catch
            {
            }
        }
    }

    private static bool CanDeleteArtifactDirectory(string directory)
    {
        if (File.Exists(Path.Combine(directory, CompletionResultFileName)))
            return true;

        if (!File.Exists(Path.Combine(directory, SubmitResultFileName)))
            return false;

        return Directory.GetCreationTimeUtc(directory) < DateTime.UtcNow.Subtract(TerminalSubmitGracePeriod);
    }

    public void Dispose()
    {
        _cleanupTimer?.Dispose();
    }
}

File: lib/EasyInk.Electron/src/main/printer/services/print-debug-log-service.ts
import { createHash } from 'crypto'
import {
  appendFileSync,
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  rmSync,
  writeFileSync
} from 'fs'
import { basename, join } from 'path'
import type { PrintRequestParams, PrinterResult } from '../../engine/models'
import type { LogLevel } from '../../engine/services/logger'
import type { HostConfig } from '../config/host-config'
import { resolvePrintDebugArtifactsDir } from '../config/host-config'

const REQUEST_FILE_NAME = 'request.json'
const INPUT_PDF_FILE_NAME = 'input.pdf'
const INPUT_HTML_FILE_NAME = 'input.html'
const MANIFEST_FILE_NAME = 'manifest.json'
const ENGINE_LOG_FILE_NAME = 'engine.log'
const SUBMIT_RESULT_FILE_NAME = 'submit-result.json'
const COMPLETION_RESULT_FILE_NAME = 'completion-result.json'

export class PrintDebugLogService {
  private readonly enabled: boolean
  private readonly artifactRoot: string
  private readonly retentionCount: number
  private readonly artifactPaths = new Map<string, string>()
  private cleanupTimer?: NodeJS.Timeout

  constructor(config: HostConfig) {
    this.enabled = config.printDebugLoggingEnabled
    this.artifactRoot = resolvePrintDebugArtifactsDir(config.printDebugArtifactsDir)
    this.retentionCount = Math.max(1, config.printDebugArtifactRetentionCount)

    if (!this.enabled) {
      return
    }

    mkdirSync(this.artifactRoot, { recursive: true })
    this.cleanupExpiredArtifacts()
    this.cleanupTimer = setInterval(() => this.cleanupExpiredArtifacts(), 24 * 60 * 60 * 1000)
    this.cleanupTimer.unref()
  }

  beginPrintRequest(requestId: string, command: string, params: PrintRequestParams): void {
    if (!this.enabled || !requestId) {
      return
    }

    const directory = this.createArtifactDirectory(requestId)
    this.artifactPaths.set(requestId, directory)
    writeJson(join(directory, REQUEST_FILE_NAME), {
      createdAt: new Date().toISOString(),
      command,
      parameters: sanitizePrintRequest(params, directory)
    })
    this.writeManifest(directory, requestId, command)
    this.cleanupExpiredArtifacts()
  }

  writeSubmitResult(requestId: string, result: PrinterResult): void {
    this.writeResult(requestId, SUBMIT_RESULT_FILE_NAME, result)
  }

  writeCompletionResult(
    requestId: string,
    request: PrintRequestParams,
    result: PrinterResult
  ): void {
    if (!this.enabled) {
      return
    }

    const directory = this.getArtifactDirectory(requestId)
    if (!directory) {
      return
    }

    writeJson(join(directory, COMPLETION_RESULT_FILE_NAME), {
      completedAt: new Date().toISOString(),
      request: summarizePrintRequest(request),
      result
    })
  }

  appendEngineLog(jobId: string | undefined, level: LogLevel, message: string): void {
    if (!this.enabled || !jobId) {
      return
    }

    const directory = this.getArtifactDirectory(jobId)
    if (!directory) {
      return
    }

    appendFileSync(
      join(directory, ENGINE_LOG_FILE_NAME),
      `[${new Date().toISOString()}] [${level.toUpperCase()}] ${message}\n`
    )
  }

  dispose(): void {
    if (this.cleanupTimer) {
      clearInterval(this.cleanupTimer)
      this.cleanupTimer = undefined
    }
  }

  private writeResult(requestId: string, fileName: string, result: PrinterResult): void {
    if (!this.enabled) {
      return
    }

    const directory = this.getArtifactDirectory(requestId)
    if (!directory) {
      return
    }

    writeJson(join(directory, fileName), {
      writtenAt: new Date().toISOString(),
      result
    })
  }

  private createArtifactDirectory(requestId: string): string {
    const dayDirectory = join(this.artifactRoot, formatDate(new Date()))
    mkdirSync(dayDirectory, { recursive: true })
    const directory = join(
      dayDirectory,
      `${formatArtifactTimestamp(new Date())}_${sanitizeFileName(requestId)}`
    )
    mkdirSync(directory, { recursive: true })
    return directory
  }

  private getArtifactDirectory(requestId: string): string | undefined {
    const cached = this.artifactPaths.get(requestId)
    if (cached && existsSync(cached)) {
      return cached
    }

    const suffix = `_${sanitizeFileName(requestId)}`
    const found = collectArtifactDirectories(this.artifactRoot)
      .filter((directory) => basename(directory).endsWith(suffix))
      .sort()
      .at(-1)

    if (found) {
      this.artifactPaths.set(requestId, found)
    }
    return found
  }

  private writeManifest(directory: string, requestId: string, command: string): void {
    writeJson(join(directory, MANIFEST_FILE_NAME), {
      requestId,
      command,
      createdAt: new Date().toISOString(),
      path: directory
    })
  }

  private cleanupExpiredArtifacts(): void {
    const artifactDirectories = collectArtifactDirectories(this.artifactRoot).sort().reverse()
    for (const directory of artifactDirectories.slice(this.retentionCount)) {
      if (canDeleteArtifactDirectory(directory)) {
        rmSync(directory, { recursive: true, force: true })
      }
    }

    if (!existsSync(this.artifactRoot)) {
      return
    }

    for (const entry of readdirSync(this.artifactRoot, { withFileTypes: true })) {
      if (!entry.isDirectory()) {
        continue
      }
      const dayDirectory = join(this.artifactRoot, entry.name)
      if (readdirSync(dayDirectory).length === 0) {
        rmSync(dayDirectory, { recursive: true, force: true })
      }
    }
  }
}

function sanitizePrintRequest(
  request: PrintRequestParams,
  directory: string
): Record<string, unknown> {
  const payload = clonePrintRequest(request) as Record<string, unknown>
  redactApiKeys(payload)

  if (typeof payload.pdfBase64 === 'string') {
    const bytes = decodeBase64(payload.pdfBase64)
    if (bytes) {
      writeFileSync(join(directory, INPUT_PDF_FILE_NAME), bytes)
      payload.pdfBase64 = describeBytes(INPUT_PDF_FILE_NAME, bytes)
    } else {
      payload.pdfBase64 = { invalidBase64: true, length: payload.pdfBase64.length }
    }
  }

  if (payload.pdfBytes) {
    const bytes = Array.isArray(payload.pdfBytes)
      ? Buffer.from(payload.pdfBytes as number[])
      : Buffer.from(payload.pdfBytes as Uint8Array)
    writeFileSync(join(directory, INPUT_PDF_FILE_NAME), bytes)
    payload.pdfBytes = describeBytes(INPUT_PDF_FILE_NAME, bytes)
  }

  for (const key of ['html', 'htmlBase64'] as const) {
    if (typeof payload[key] === 'string') {
      const html = key === 'htmlBase64' ? decodeBase64Text(payload[key]) : payload[key]
      if (html != null) {
        writeFileSync(join(directory, INPUT_HTML_FILE_NAME), html, 'utf8')
        payload[key] = { savedAs: INPUT_HTML_FILE_NAME, chars: html.length }
      }
    }
  }

  if (payload.viewer && typeof payload.viewer === 'object') {
    const viewer = payload.viewer as { pages?: unknown[] }
    payload.viewer = {
      ...viewer,
      pages: Array.isArray(viewer.pages)
        ? { count: viewer.pages.length, chars: viewer.pages.map(String).join('').length }
        : viewer.pages
    }
  }

  return payload
}

function clonePrintRequest(value: unknown): unknown {
  if (Buffer.isBuffer(value)) {
    return Buffer.from(value)
  }
  if (value instanceof Uint8Array) {
    return new Uint8Array(value)
  }
  if (Array.isArray(value)) {
    return value.map(clonePrintRequest)
  }
  if (!value || typeof value !== 'object') {
    return value
  }

  return Object.fromEntries(
    Object.entries(value as Record<string, unknown>).map(([key, child]) => [
      key,
      clonePrintRequest(child)
    ])
  )
}

function summarizePrintRequest(request: PrintRequestParams): Record<string, unknown> {
  return {
    printerName: request.printerName,
    copies: request.copies,
    dpi: request.dpi,
    landscape: request.landscape,
    paperSize: request.paperSize,
    offset: request.offset,
    userData: request.userData,
    hasPdfBytes: Boolean(request.pdfBytes),
    hasPdfBase64: Boolean(request.pdfBase64),
    hasPdfUrl: Boolean(request.pdfUrl),
    hasHtml: request.html != null,
    hasHtmlBase64: Boolean(request.htmlBase64),
    hasHtmlUrl: Boolean(request.htmlUrl),
    hasViewer: Boolean(request.viewer)
  }
}

function redactApiKeys(value: unknown): void {
  if (Array.isArray(value)) {
    value.forEach(redactApiKeys)
    return
  }

  if (!value || typeof value !== 'object') {
    return
  }

  for (const [key, child] of Object.entries(value)) {
    if (key.toLowerCase() === 'apikey' || key.toLowerCase() === 'x-api-key') {
      ;(value as Record<string, unknown>)[key] = '***'
    } else {
      redactApiKeys(child)
    }
  }
}

function describeBytes(savedAs: string, bytes: Buffer): Record<string, unknown> {
  return {
    savedAs,
    bytes: bytes.length,
    sha256: createHash('sha256').update(bytes).digest('hex')
  }
}

function decodeBase64(value: string): Buffer | undefined {
  try {
    return Buffer.from(stripDataUrlPrefix(value), 'base64')
  } catch {
    return undefined
  }
}

function decodeBase64Text(value: string): string | undefined {
  const bytes = decodeBase64(value)
  return bytes ? bytes.toString('utf8') : undefined
}

function stripDataUrlPrefix(value: string): string {
  const marker = ';base64,'
  const index = value.indexOf(marker)
  return index >= 0 ? value.slice(index + marker.length) : value
}

function sanitizeFileName(value: string): string {
  const sanitized = [...value]
    .map((char) => (isInvalidFileNameChar(char) ? '_' : char))
    .join('')
    .trim()
  return sanitized || 'request'
}

function isInvalidFileNameChar(char: string): boolean {
  return '<>:"/\\|?*'.includes(char) || char.charCodeAt(0) < 32
}

function collectArtifactDirectories(root: string): string[] {
  if (!existsSync(root)) {
    return []
  }

  const directories: string[] = []
  for (const day of readdirSync(root, { withFileTypes: true })) {
    if (!day.isDirectory()) {
      continue
    }
    const dayDirectory = join(root, day.name)
    for (const item of readdirSync(dayDirectory, { withFileTypes: true })) {
      const directory = join(dayDirectory, item.name)
      if (item.isDirectory() && existsSync(join(directory, MANIFEST_FILE_NAME))) {
        directories.push(directory)
      }
    }
  }
  return directories
}

function canDeleteArtifactDirectory(directory: string): boolean {
  if (existsSync(join(directory, COMPLETION_RESULT_FILE_NAME))) {
    return true
  }

  if (!existsSync(join(directory, SUBMIT_RESULT_FILE_NAME))) {
    return false
  }

  try {
    const manifest = JSON.parse(readFileSync(join(directory, MANIFEST_FILE_NAME), 'utf8')) as {
      createdAt?: string
    }
    return manifest.createdAt
      ? Date.now() - new Date(manifest.createdAt).getTime() > 60 * 60 * 1000
      : true
  } catch {
    return false
  }
}

function writeJson(path: string, value: unknown): void {
  writeFileSync(path, `${JSON.stringify(value, null, 2)}\n`, 'utf8')
}

function formatDate(date: Date): string {
  return `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}-${String(
    date.getDate()
  ).padStart(2, '0')}`
}

function formatArtifactTimestamp(date: Date): string {
  return `${date.getFullYear()}${String(date.getMonth() + 1).padStart(2, '0')}${String(
    date.getDate()
  ).padStart(2, '0')}-${String(date.getHours()).padStart(2, '0')}${String(
    date.getMinutes()
  ).padStart(2, '0')}${String(date.getSeconds()).padStart(2, '0')}${String(
    date.getMilliseconds()
  ).padStart(3, '0')}`
}
